.github/workflows/tests.yml

```diff
@@ -35,12 +35,54 @@ jobs:
       - name: Coverage reporting
         run: |
           coverage combine
-          coverage report -m
-          coverage xml
-          coverage html
-      - name: Publish coverage results
-        uses: codecov/codecov-action@v4
+          coverage report --show-missing
+          mv .coverage .coverage.${{ matrix.python-version }}
+      - name: Store coverage file
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage-${{ matrix.python-version }}
+          path: .coverage.${{ matrix.python-version }}
+
+  coverage:
+    name: Coverage
+    runs-on: ubuntu-latest
+    needs: tests
+    permissions:
+      # If the author is a maintainer, the permission level is set by the
+      # values below.
+      # `pull-requests: write` is needed for publishing new comments in pull
+      # requests.
+      # `contents: write` is needed for pushing data to the
+      # `python-coverage-comment-action` branch, and for editing existing
+      # comments (to avoid publishing multiple comments in the same PR)
+      # In case the pull request comes from a forked repository, the maximum
+      # permission level is read, so the permissions below won't be acted upon
+      # by GitHub.
+      # https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token
```

Comment:
I've tried being a bit verbose because it's important that readers understand the security implications on those lines.

If you prefer without the comments or with less comments, let me know.